### ISSUE TITLE
Do you plan to incorporate a notion of frames into the manif library?

### ISSUE THREAD

**Issue by adas-dev** [artivis/manif#215]
Dear manif-library creators,

Do you have any plans to implement compile-time coordinate frame semantics checking as it is done in the project https://github.com/wavelab/wave_geometry?

I was pondered by an idea of compile time coordinate frame checking for quite a while and even written some prototype code. But then I have found the mention of `wave_geometry` in one of your papers: https://www.theoj.org/joss-papers/joss.01371/10.21105.joss.01371.pdf.

I used neither `manif` nor `wave_geometry` so far in my work but I plan to migrate to `manif` because of numerous obvious benefits that it provides. It would also be nice to have a compile- and run-time coordinate frame checking in `manif` one day.

**Comment by artivis** [comment]
Hi,

Altho interesting, there is no plan to develop such feature in `manif`. I have a limited bandwidth for this project and there are other priorities I'd like to get done. This being said, I can provide *some* support shall you want to take a shot at implementing it.

> I plan to migrate to `manif`

May I ask, from which library would you be migrating?

**Comment by adas-dev** [comment]
I mostly used an OpenVINS (https://github.com/rpng/open_vins/) codebase for my research. The pure Eigen types are used there, so *I am planning to migrate from Eigen :-)*. 

Not knowing much about the existence of specific well-tested libraries out there interoperable with pure Eigen types I started implementing my own classes that derive from Eigen types and represent spacial primitives. The goal was to eliminate the need to remember and manipulate indices while accessing individual dimensions of some lengthy state vector. It turned out that just adding interfaces and some semantics to raw Eigen types is a hard and laborious task which also requires implementing lots of unit tests. Since I am interested in a more high-level algorithms I started looking for what is available in the open-source community. 
(To be honest I knew about the existence of the `manif` a long time ago but at that time I thought that migrating the whole OpenVINS to `manif` would be too time consuming.)

I am interested in incorporating the notion of frames into the `manif` but before that I need to learn about the library and the way it is implemented. I think it will take a lot of time before I will have a rough understanding of how it can be done in the case of this library. But it is very nice to know that you are interested in providing some support regarding this topic. The active support and maintenance of this library from authors is one of the main factors contributing to my decision to migrate to it. (The other two factors are the test coverage and the paper.)

Thank you for the answer. I am closing this task since my question is answered.
